[otbn,doc] Documentation improvements (#17165)
lowRISC/opentitan: hw/ip/otbn/dv/otbnsim/README.md

# OTBNSim - OTBN Simulation Software

> `OTBNSim` class represents the ISS in a traditional sense.

Reviewer: What does 'traditional' mean here? Can you elaborate?

> It has methods for starting the executing, cycle-accurate simulation tracing and also abilty to simulate injected errors and wiping as a response to those errors.

Reviewer: ```suggestion
It has methods for starting the execution, for cycle-accurate simulation tracing, and for injecting errors and react to those errors with a secure wipe.
```

As explained in the overall [OTBN DV documentation](../../doc/dv/_index.md), this simulator is driven essentially from the SystemVerilog model of the OTBN module.

> Simulator works in a step-by-step fashion and also has multiple different methods to mimic external stimulus to OTBN.

Reviewer: ```suggestion
The simulator works in a step-by-step fashion and it has multiple methods to apply external stimuli to OTBN.
```

> At a typical run without errors ISS does the following:

Reviewer: ```suggestion
In a typical run without errors, the ISS does the following:
```

>  1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`/

Reviewer: ```suggestion
 1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`.
```

>  2. Loading decoded program to a local list in `sim.py`.

Reviewer: ```suggestion
 2. Load the decoded program to a local list in `sim.py`.
```

 3. With each `step` command from the SystemVerilog side, update simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).
 4. Once the step is done, pass the generated trace to `iss_wrapper.cc` to then get it passed to `OTBNTraceChecker`.

## Co-Simulation with RTL

> Co-simulation works by using a tracer module (`otbn_tracer`) that spits out some logs from the RTL and having a serialised result in a text form on `stdout` to produce same style of logs from the Python model.

Reviewer: ```suggestion
For co-simulation of RTL and ISS, the `otbn_tracer` module logs state changes of the RTL, and the ISS logs state changes of the Python model.
```
I don't think the `stdout` mechanics should be in the first sentence explaining co-simulation, so I removed it. If you think it should be mentioned at this point, I suggest adding a second sentence.

> These two seperate logs generated by the model and the RTL gets compared for checking correct behaviour.

Reviewer: ```suggestion
These two separate logs generated by the model and the RTL get compared.
If the state of the RTL ever differs from the state of the model, an error is reported.
```

> For more information about how OTBN RTL produces traces see. [Tracer README](../tracer/README.md).

Reviewer: ```suggestion
For more information about how OTBN RTL produces traces see the [Tracer README](../tracer/README.md).
```

# OTBNSim - OTBN Simulation Software

> `OTBNSim` class represents the simulator that runs a model of the OTBN block as a whole.

Reviewer 2: ```suggestion
`OTBNSim` class represents a simulator that models of the whole OTBN block.
```

> Therefore it includes abstract models for the the state of the core, memory modules of it and also the registers.

Reviewer 2: ```suggestion
Therefore it includes abstract models for the the state of the core, its memory modules, and also the register file.
```

> It has methods for starting the execution, for cycle-accurate simulation tracing, and for injecting errors and react to those errors with a secure wipe.

Reviewer 2: ```suggestion
It has methods for starting the execution, cycle-accurate simulation tracing, and for injecting errors and reacting to those errors with a secure wipe.
```

As explained in the overall [OTBN DV documentation](../../doc/dv/_index.md), this simulator is driven essentially from the SystemVerilog model of the OTBN module.

The simulator works in a step-by-step fashion and it has multiple methods to apply external stimuli to OTBN.
In a typical run without errors, the ISS does the following:

>  1. Decode the program on the path given by `iss_wrapper.cc` by using `decode_file` method in `decode.py`.

Reviewer 2: ```suggestion
 1. Decode the program on the path given by `iss_wrapper.cc` by using the `decode_file` method in `decode.py`.
```

 2. Load the decoded program to a local list in `sim.py`.

>  3. With each `step` command from the SystemVerilog side, update simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).

Reviewer 2: ```suggestion
 3. With each `step` command from the SystemVerilog side, update the simulated state of the core (`state.py`), registers (`wsr.py`, `csr.py` and `gpr.py`) and data memory (`dmem.py`).
```

>  4. Once the step is done, pass the generated trace to `iss_wrapper.cc` to then get it passed to `OTBNTraceChecker`.

Reviewer 2: ```suggestion
 4. Once the step is done, pass the generated trace to `iss_wrapper.cc`, which then passes it on to `OTBNTraceChecker`.
```

## Co-Simulation with RTL
For co-simulation of RTL and ISS, the `otbn_tracer` module logs state changes of the RTL, and the ISS logs state changes of the Python model.
Trace entries from the simulated core (aka. from RTL) appear as a result of DPI callbacks while ISS trace entries appear in the trace checker through `ISSWrapper` using `OnIssTrace` method after sending a step command to `OTBNSim`.

> These two separate logs generated by the model and the RTL get compared for checking correct behaviour.

Reviewer 2: ```suggestion
 To check for correct behaviour, the two separate logs generated by the model and the RTL are compared.
```

For more information about how OTBN RTL produces traces see the [Tracer README](../tracer/README.md).

> To see C++ program that compares both traces check `otbn_trace_checker.cc` on `../model/otbn_trace_entry`.

Reviewer 2: ```suggestion
To see the C++ program that compares both traces, check the method `otbn_trace_checker.cc` in `../model/otbn_trace_entry`.
```
